File: backend/app/services/french_tax_ids.py
"""
Validation des identifiants d'entreprise français : SIREN, SIRET, n° de TVA intracommunautaire.

Références (algorithmes publics, non propriétaires) :
- SIREN/SIRET : clé de Luhn (norme INSEE), à l'exception des établissements
  du groupe La Poste (SIREN 356000000) qui suivent une règle particulière
  (somme des chiffres multiple de 5) — gérée séparément ci-dessous.
- TVA intracommunautaire FR : clé = (12 + 3 × (SIREN mod 97)) mod 97, sur 2 chiffres.
  Les numéros de TVA à clé alphabétique (ancien schéma, hors formule numérique)
  ne sont pas vérifiables par ce calcul : on ne les rejette pas, on signale
  simplement que la clé n'a pas pu être confirmée.
"""
import re
from dataclasses import dataclass

LA_POSTE_SIREN = "356000000"


def _luhn_ok(digits: str) -> bool:
    total = 0
    for i, ch in enumerate(reversed(digits)):
        d = int(ch)
        if i % 2 == 1:
            d *= 2
            if d > 9:
                d -= 9
        total += d
    return total % 10 == 0


def clean_digits(value: str) -> str:
    return re.sub(r"[^0-9]", "", value or "")


def is_valid_siren(value: str) -> bool:
    digits = clean_digits(value)
    if len(digits) != 9:
        return False
    return _luhn_ok(digits)


def is_valid_siret(value: str) -> bool:
    digits = clean_digits(value)
    if len(digits) != 14:
        return False
    siren, nic = digits[:9], digits[9:]
    if siren == LA_POSTE_SIREN:
        return sum(int(c) for c in digits) % 5 == 0
    return _luhn_ok(digits)


def compute_tva_key(siren: str) -> str | None:
    """Clé de contrôle à 2 chiffres selon la formule officielle. None si le SIREN est invalide."""
    if not is_valid_siren(siren):
        return None
    key = (12 + 3 * (int(siren) % 97)) % 97
    return f"{key:02d}"


@dataclass
class TvaCheck:
    valid: bool | None  # True/False si vérifiable, None si le format échappe à la formule numérique
    siren: str | None
    message: str


def validate_tva_fr(value: str) -> TvaCheck:
    cleaned = re.sub(r"[\s.-]", "", (value or "").upper())
    m = re.match(r"^FR([0-9A-Z]{2})(\d{9})$", cleaned)
    if not m:
        return TvaCheck(valid=False, siren=None, message="Format attendu : FR + 2 caractères + 9 chiffres (SIREN).")

    key, siren = m.group(1), m.group(2)
    if not is_valid_siren(siren):
        return TvaCheck(valid=False, siren=siren, message="Le SIREN contenu dans le n° de TVA est invalide.")

    if key.isdigit():
        expected = compute_tva_key(siren)
        if key == expected:
            return TvaCheck(valid=True, siren=siren, message="N° de TVA valide.")
        return TvaCheck(valid=False, siren=siren, message=f"Clé de contrôle incorrecte (attendu {expected}).")

    # Ancien schéma alphabétique — le SIREN est valide mais la clé n'est pas vérifiable ici.
    return TvaCheck(valid=None, siren=siren, message="SIREN valide, mais clé alphabétique non vérifiable automatiquement.")


def identify_and_check(value: str) -> str | None:
    """
    Détecte s'il s'agit d'un SIRET, SIREN ou n° de TVA FR à partir de sa forme,
    et retourne un message d'avertissement si le contrôle échoue — None si tout va bien
    ou si la forme ne correspond à aucun de ces identifiants (on ne force rien).
    """
    if not value:
        return None
    stripped = value.strip().upper()

    if stripped.startswith("FR") and any(c.isdigit() for c in stripped):
        check = validate_tva_fr(stripped)
        if check.valid is False:
            return f"N° de TVA « {value} » invalide : {check.message}"
        return None

    digits = clean_digits(stripped)
    if len(digits) == 14:
        return None if is_valid_siret(digits) else f"SIRET « {value} » invalide (clé de contrôle incorrecte)."
    if len(digits) == 9:
        return None if is_valid_siren(digits) else f"SIREN « {value} » invalide (clé de contrôle incorrecte)."

    return None  # ni SIRET, ni SIREN, ni TVA FR reconnaissable — on ne le signale pas comme une erreur

File: frontend/lib/frenchTaxIds.ts
/**
 * Portage JS des contrôles SIREN/SIRET/TVA FR — même algorithme que
 * backend/app/services/french_tax_ids.py, pour un retour immédiat dans le
 * formulaire sans aller-retour serveur. Le backend reste la source de vérité
 * (revalidé à l'extraction et à la sauvegarde).
 */
const LA_POSTE_SIREN = "356000000";

function luhnOk(digits: string): boolean {
  let total = 0;
  const rev = digits.split("").reverse();
  rev.forEach((ch, i) => {
    let d = Number(ch);
    if (i % 2 === 1) {
      d *= 2;
      if (d > 9) d -= 9;
    }
    total += d;
  });
  return total % 10 === 0;
}

function cleanDigits(value: string): string {
  return (value || "").replace(/[^0-9]/g, "");
}

export function isValidSiren(value: string): boolean {
  const digits = cleanDigits(value);
  if (digits.length !== 9) return false;
  return luhnOk(digits);
}

export function isValidSiret(value: string): boolean {
  const digits = cleanDigits(value);
  if (digits.length !== 14) return false;
  const siren = digits.slice(0, 9);
  if (siren === LA_POSTE_SIREN) {
    const sum = digits.split("").reduce((acc, c) => acc + Number(c), 0);
    return sum % 5 === 0;
  }
  return luhnOk(digits);
}

export function computeTvaKey(siren: string): string | null {
  if (!isValidSiren(siren)) return null;
  const key = (12 + 3 * (Number(siren) % 97)) % 97;
  return String(key).padStart(2, "0");
}

export function validateTvaFr(value: string): { valid: boolean | null; message: string } {
  const cleaned = (value || "").toUpperCase().replace(/[\s.-]/g, "");
  const m = cleaned.match(/^FR([0-9A-Z]{2})(\d{9})$/);
  if (!m) return { valid: false, message: "Format attendu : FR + 2 caractères + 9 chiffres (SIREN)." };

  const [, key, siren] = m;
  if (!isValidSiren(siren)) return { valid: false, message: "Le SIREN contenu dans le n° de TVA est invalide." };

  if (/^\d{2}$/.test(key)) {
    const expected = computeTvaKey(siren);
    if (key === expected) return { valid: true, message: "N° de TVA valide." };
    return { valid: false, message: `Clé de contrôle incorrecte (attendu ${expected}).` };
  }
  return { valid: null, message: "SIREN valide, clé alphabétique non vérifiable automatiquement." };
}

/** Message d'avertissement si la valeur ressemble à un SIRET/SIREN/TVA FR invalide, sinon null. */
export function identifyAndCheck(value: string): string | null {
  if (!value) return null;
  const stripped = value.trim().toUpperCase();

  if (stripped.startsWith("FR") && /\d/.test(stripped)) {
    const check = validateTvaFr(stripped);
    return check.valid === false ? `N° de TVA invalide : ${check.message}` : null;
  }

  const digits = cleanDigits(stripped);
  if (digits.length === 14) return isValidSiret(digits) ? null : "SIRET invalide (clé de contrôle incorrecte).";
  if (digits.length === 9) return isValidSiren(digits) ? null : "SIREN invalide (clé de contrôle incorrecte).";
  return null;
}
